src/Workspaces/Core/Portable/Classification/SyntaxClassification/SyntacticChangeRangeComputer.cs

```diff
@@ -0,0 +1,249 @@
+    /// <remarks>
+    /// This computation is not guaranteed to be minimal.  It may return a range that includes parts that are unchanged.
+    /// This means it is also legal for the change range to just specify the entire file was changed. The quality of
+    /// results will depend on how well the parsers did with incremental parsing, and how much time is given to do the
+    /// comparison.  In practice, for large files (i.e. 15kloc) with standard types of edits, this generally returns
+    /// results in around 50-100 usecs on a i7 3GHz desktop.
+    /// <para>
+    /// This algorithm will respect the timeout provided to the best of abilities.  If any information has been computed
+    /// when the timeout elapses, it will be returned.
+    /// </para>
+    /// </remarks>
+    internal static class SyntacticChangeRangeComputer
+    {
+        private static readonly ObjectPool<Stack<SyntaxNodeOrToken>> s_pool = new(() => new());
+
+        public static async ValueTask<TextChangeRange?> ComputeSyntacticChangeRangeAsync(
+            Document oldDocument, Document newDocument, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            // If they're the same doc, there is no change.
+            if (oldDocument == newDocument)
+                return new TextChangeRange();
+
+            var stopwatch = SharedStopwatch.StartNew();
+
+            var oldRoot = await oldDocument.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // If we ran out of time, we have to assume both are completely different.
+            if (stopwatch.Elapsed > timeout)
+                return null;
+
+            var newRoot = await newDocument.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            if (stopwatch.Elapsed > timeout)
+                return null;
+
+            return ComputeSyntacticChangeRange(oldRoot, newRoot, timeout, stopwatch, cancellationToken);
+        }
+
+        public static TextChangeRange ComputeSyntacticChangeRange(SyntaxNode oldRoot, SyntaxNode newRoot, TimeSpan timeout, CancellationToken cancellationToken)
+            => ComputeSyntacticChangeRange(oldRoot, newRoot, timeout, SharedStopwatch.StartNew(), cancellationToken);
+
+        private static TextChangeRange ComputeSyntacticChangeRange(
+            SyntaxNode oldRoot, SyntaxNode newRoot, TimeSpan timeout, SharedStopwatch stopwatch, CancellationToken cancellationToken)
+        {
+            if (oldRoot == newRoot)
+                return default;
+
+            using var leftOldStack = s_pool.GetPooledObject();
+            using var leftNewStack = s_pool.GetPooledObject();
+            using var rightOldStack = s_pool.GetPooledObject();
+            using var rightNewStack = s_pool.GetPooledObject();
+
+            leftOldStack.Object.Push(oldRoot);
+            leftNewStack.Object.Push(newRoot);
+            rightOldStack.Object.Push(oldRoot);
+            rightNewStack.Object.Push(newRoot);
```

Comment:
The stacks are only used in the implementations of the local methods? Just move them in? That way we're not allocating ones which may not be used (the right stacks), and it probably also means the local functions can become static?